llvm/lib/Transforms/Utils/SCCPSolver.cpp

```diff
@@ -157,6 +162,15 @@ static bool refineInstruction(SCCPSolver &Solver,
                           GEPNoWrapFlags::noUnsignedWrap());
       Changed = true;
     }
+  } else if (match(&Inst, m_And(m_Value(X), m_LowBitMask(RHSC)))) {
+    ConstantRange LRange = GetRange(Inst.getOperand(0));
+    if (!LRange.getUnsignedMax().ule(*RHSC))
+      return false;
+
+    Inst.replaceAllUsesWith(X);
+    Inst.eraseFromParent();
+    ++InstRemovedStat;
+    Changed = true;
```

Comment:
This doesn't really fit with the other "refine" transforms, which all infer flags. I'd probably handle it outside this function, but no strong opinion.